File: 3° Período/Estrutura de Dados I/Atividade 12/model/FilaEncadeada.js
class FilaEncadeada {
  #inicio;
  #fim;
  #qtd;

  constructor() {
    this.#inicio = null;
    this.#fim = null;
    this.#qtd = 0;
  }

  isEmpty() {
    return this.#inicio === null;
  }

  enqueue(novoElemento) {
    const novoNo = new No(novoElemento);

    if (this.isEmpty()) {
      this.#inicio = novoNo;
    } else {
      this.#fim.proximo = novoNo;
    }

    this.#fim = novoNo;
    this.#qtd++;

    console.log(`Adicionado:${novoElemento}`);
    console.log(`Quantidade:${this.#qtd}`);

    return true;
  }

  dequeue() {
    if (!this.isEmpty()) {
      let removido = this.#inicio.dado;

      this.#inicio = this.#inicio.proximo;
      this.#qtd--;

      console.log(`Removido: ${removido}`);
      console.log(`Quantidade: ${this.#qtd}`);

      return removido;
    }
    return null;
  }

  toString() {
    let resultado = "";
    let aux = this.#inicio;

    while (aux !== null) {
      resultado += aux.dado + " | ";
      aux = aux.proximo;
    }

    return resultado;
  }

  [Symbol.iterator]() {
    let aux = this.#inicio;

    return {
      next() {
        if (aux !== null) {
          const value = aux.dado;
          aux = aux.proximo;
          return { value, done: false };
        } else {
          return { done: true };
        }
      },
    };
  }
}

File: 3° Período/EstruturaDeDados_I/Lista01/Codigo16/src/FilaEncadeada.java
import java.util.Iterator;

public class FilaEncadeada implements Iterable<Integer> {
    private No inicio;
    private No fim;
    private int qtd;

    public FilaEncadeada() {
        this.inicio = null;
        this.fim = null;
        this.qtd = 0;
    }

    public boolean isEmpty() {
        return this.inicio == null;
    }

    public boolean enqueue(int novoElemento) {
        No novoNo = new No(novoElemento);

        if (this.isEmpty()) {
            this.inicio = novoNo;
        } else {
            this.fim.setProximo(novoNo);
        }

        this.fim = novoNo;
        this.qtd++;

        return true;
    }

    public Integer dequeue() {
        if (!this.isEmpty()) {
            int removido = this.inicio.getDado();

            this.inicio = this.inicio.getProximo();
            this.qtd--;

            return removido;
        }
        return null;
    }

    public int posicaoQueElementoSeEncontraNaFila(int elemento) {
        int posicao = 1;

        for (Integer e : this) {
            if (elemento == e) {
                return posicao;
            }
            posicao++;
        }

        return -1;
    }

    @Override
    public String toString() {
        String resultado = "";
        No aux = this.inicio;

        while (aux != null) {
            resultado += aux.getDado() + " | ";
            aux = aux.getProximo();
        }

        return resultado;
    }

    @Override
    public Iterator<Integer> iterator() {
        return new Iterator<Integer>() {
            private No aux = FilaEncadeada.this.inicio;

            @Override
            public boolean hasNext() {
                return aux != null;
            }

            @Override
            public Integer next() {
                if (aux != null) {
                    int value = aux.getDado();
                    aux = aux.getProximo();
                    return value;
                }
                return null;
            }
        };
    }
}
